src/D2L.CodeStyle.Analyzers/Async/Generator/SyntaxTransformer.cs

```diff
@@ -0,0 +1,120 @@
+﻿#nullable enable
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace D2L.CodeStyle.Analyzers.Async.Generator;
+
+/// <summary>
+/// Base class for things that transform a SyntaxNode into another and collect
+/// diagnostics. This is useful for writing generators that copy+paste code
+/// with only light modifications.
+/// </summary>
+public abstract class SyntaxTransformer {

```

Comment:
I don't actually anticipate a second one of these (I hope) but I was keeping the common stuff in a separate class I could drop the `private`/`protected` stuff (which I'm using to hide like, `m_diagnostics` behind a dinky API) and use a `partial` if you dislike it.

Reply:
To use this class you `new` up a transformer and hand it a node, and it will either give you a node back or a bunch of diagnostics.